### DISCUSSION TITLE
How to custom favicon docs?

### DISCUSSION THREAD

**Discussion by letanloc1998** [fastapi/fastapi#6936] [Questions] [answered]
### First check

* [ ] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

I want to self-host all resource but I can't find any document change logo swagger
I only see https://fastapi.tiangolo.com/advanced/extending-openapi/ to host js and css

I want to change `<link rel="shortcut icon" href="https://fastapi.tiangolo.com/img/favicon.png">` store in my host

Moreover, does anyone help me link to custom template,... swagger and redoc? Thanks

### Additional context
http://localhost:8000/docs

**Comment by ycd** [comment, marked as answer]
You can override the default `redoc_url` and `docs_url`, then you can create your own endpoint. Which you can customize however if you want.


```python
from fastapi import FastAPI
from fastapi.openapi.docs import get_swagger_ui_html, get_redoc_html


app = FastAPI(docs_url=None, redoc_url=None)


@app.get("/docs", include_in_schema=False)
def overridden_swagger():
	return get_swagger_ui_html(openapi_url="/openapi.json", title="FastAPI", swagger_favicon_url="https://placekitten.com/200/300")

@app.get("/redoc", include_in_schema=False)
def overridden_redoc():
	return get_redoc_html(openapi_url="/openapi.json", title="FastAPI", redoc_favicon_url="https://placekitten.com/200/300")
```

**Comment by letanloc1998** [comment]
Thank you a lot
I has been override but I didn't know "redoc_favicon_url" before. I will do it

**Comment by tiangolo** [comment]
Thanks for the help here @ycd ! 👏  🙇 

Thanks for reporting back and closing the issue @letanloc1998  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by brunolnetto** [comment]
How can I refer to openapi favicon.ico placed on my static folder like below?

```
from fastapi import FastAPI

app = FastAPI()

obj=StaticFiles(directory="static")
app.mount("/static", obj, name="static")
```

**Comment by clawfrank** [comment]
@brunolnetto  Here's an example with a local favicon.ico

    
    app.mount("/static", StaticFiles(directory="/app/static"), name="static")
    
    
    # Have to override these so the swagger/redoc pages don't pull from a CDN
    
    @app.get("/docs", include_in_schema=False)
    async def custom_swagger_ui_html():
        return get_swagger_ui_html(
            openapi_url=app.openapi_url,
            title=app.title + " - Swagger UI",
            oauth2_redirect_url=app.swagger_ui_oauth2_redirect_url,
            swagger_js_url="/static/swagger-ui-bundle.js",
            swagger_css_url="/static/swagger-ui.css",
            swagger_favicon_url="/static/fastapi.png"
        )

This doc page could use updating to include this: https://fastapi.tiangolo.com/how-to/custom-docs-ui-assets/
